subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/ForkingGradleSession.groovy

```diff
@@ -0,0 +1,106 @@
+ * for low overhead.
+ */
+@CompileStatic
+class ForkingGradleSession implements GradleSession {
+
+    final GradleInvocationSpec invocation
+    private final IntegrationTestBuildContext integrationTestBuildContext
+
+    private ProcessBuilder stop
+
+    ForkingGradleSession(GradleInvocationSpec invocation, IntegrationTestBuildContext integrationTestBuildContext) {
+        this.invocation = invocation
+        this.integrationTestBuildContext = integrationTestBuildContext
+    }
+
+    @Override
+    void prepare() {
+        cleanup()
+    }
+
+    Action<MeasuredOperation> runner(BuildExperimentInvocationInfo invocationInfo, InvocationCustomizer invocationCustomizer) {
+        return { MeasuredOperation measuredOperation ->
+            DurationMeasurementImpl.measure(measuredOperation) {
+                run(invocationInfo, invocationCustomizer)
+            }
+        } as Action<MeasuredOperation>
+    }
+
+    @Override
+    void cleanup() {
+        if (stop != null) {
+            stop.start().waitFor()
+            stop = null
+        }
+    }
+
+    private void run(BuildExperimentInvocationInfo invocationInfo, InvocationCustomizer invocationCustomizer) {
+        def invocation = invocationCustomizer ? invocationCustomizer.customize(invocationInfo, this.invocation) : this.invocation
+
+        List<String> args = []
+        if (OperatingSystem.current().isWindows()) {
+            args << "cmd.exe" << "/C"
+        }
+        args << new File(integrationTestBuildContext.gradleHomeDir, "bin/gradle").absolutePath
+        args << "--gradle-user-home" << new File(invocationInfo.projectDir, "gradle-user-home").absolutePath
+        args << "--no-search-upward"
+        args << "--stacktrace"
+        args << "-Dorg.gradle.jvmargs=${invocation.jvmOpts.collect { '\'' + it + '\'' }.join(' ')}".toString()
+        args += invocation.args
+        if (invocation.useDaemon) {
+            args << "--daemon"
+        } else {
+            args << "--no-daemon"
+        }
+
+        ProcessBuilder run = newProcessBuilder(invocationInfo, args + invocation.tasksToRun)
+        stop = newProcessBuilder(invocationInfo, args + "--stop")
+
+        def exitCode = run.start().waitFor()
+        if (exitCode != 0 && !invocation.expectFailure) {
+            throw new IllegalStateException("Build failed, see ${getBuildLog(invocationInfo)} for details")
+        }
+    }
+
+    private ProcessBuilder newProcessBuilder(BuildExperimentInvocationInfo invocationInfo, List<String> args) {
+        new ProcessBuilder()
+            .directory(invocationInfo.projectDir)
+            .redirectOutput(Redirect.appendTo(getBuildLog(invocationInfo)))
+            .redirectError(Redirect.appendTo(getBuildLog(invocationInfo)))
+            .command(args)
+    }
+
+    private File getBuildLog(BuildExperimentInvocationInfo invocationInfo) {
+        new File(invocationInfo.projectDir, "log.txt")
```

Comment:
We should make sure the log is preserved on teamcity (add it to artifacts in the Individual performance worker and adhoc configurations?)